_posts/2020-07-30-hhvm-4.68.markdown

```diff
@@ -0,0 +1,115 @@
+    are neither `varray` nor `darray`, depending on the exact INI options.
+  - Migrating `array` type annotations to `varray` and `darray` (or even `vec`
+    and `dict`) might be preferable, but it may cause runtime errors, possibly
+    undetected by the typechecker (for values coming from untyped code including
+    some built-ins). You can use `<<__Soft>>` to turn these errors into
+    warnings.
+  - HHAST migration for `array` type annotations should be available soon.
+- The functions `current`, `key`, `next`, `prev`, `reset`, `end` were removed
+  (arrays no longer maintain an "internal iterator", which should also reduce
+  their memory usage in the future).
+  - Additional iterators and related classes were removed:
+    - all methods of `ArrayIterator` other than `current`, `key`, `next`,
+      `rewind`, `valid`
+    - `AppendIterator`
+    - `CachingIterator`
+    - `CallbackFilterIterator`
+    - `LimitIterator`
+    - `MultipleIterator`
+    - `ParentIterator`
+    - `RecursiveArrayIterator`
+    - `RecursiveCachingIterator`
+    - `RecursiveCallbackFilterIterator`
+    - `RecursiveTreeIterator`
+    - `SplFixedArray`
+    - `SplObjectStorage`
+  - Most users of these iteration APIs can be rewritten with `foreach` loops,
+    and would be much faster to boot. For cases that can't be converted this
+    way, you can use `Vec\keys($array)` to get an array of keys and
+    `vec($array)` to get an array of values and use a `for` loop over indices
+    for these two arrays.
+- Attempting to merge arrays using the `+` or `+=` operators now throws an
+  exception. It had already been a typechecker error.
+  - In recent HHVM versions, you can use the INI options
+    `hhvm.hack_arr_compat_notices=true` and
+    `hhvm.hack_arr_compat_check_array_plus=true` to raise a notice.
+- Fixed the position at which the typechecker reports an error for a missing
+  type constant. This may require some `HH_FIXME`s to be moved.
+- Using `meth_caller` with a trait now raises a typechecker error. Runtime
+  behavior hasn't changed, HHVM currently allows constructing such `meth_caller`
+  but throws an exception when the method call is invoked.
+- By default, HHVM now raises a warning for any `switch` statements where none
+  of the cases matched the provided value. In the future, this will be changed
+  to throw an exception.
+  - A `default` branch should be added to any `switch` statements that
+    intentionally don't cover all possible cases.
+  - Use the INI option `hhvm.throw_on_non_exhaustive_switch=0` to temporarily
+    disable this warning, or `hhvm.throw_on_non_exhaustive_switch=2` to enable
+    the future behavior now (throw an exception).
+- By default, HHVM now triggers an unrecoverable error if any function return
+  type annotations are violated.
+  - The INI option
+    [`hhvm.check_return_type_hints`](https://github.com/facebook/hhvm/blob/HHVM-4.68/hphp/runtime/base/runtime-option.h#L734)
```

Comment:
I tried on the latest nightly and `function _(): int { return ''; }` throws a `TypeError` when invoked.
Is the intent to behave like `\trigger_error('Typeviolation...', \E_ERROR);` which is truly unrecoverable to throw a TypeError, which only is unrecoverable when uncaught.

Reply:
https://github.com/facebook/hhvm/blob/master/hphp/runtime/base/runtime-error.cpp#L106-L114
```CPP

void raise_return_typehint_error(const std::string& msg) {
  if (RuntimeOption::PHP7_EngineExceptions) {
    VMRegAnchor _;
    SystemLib::throwTypeErrorObject(msg);
  }
  raise_recoverable_error(msg);
  raise_error("Error handler tried to recover from a return typehint violation");
}
```

Does someone know the default value of `RuntimeOption::PHP7_EngineExceptions`?

Reply:
> The option is on by default, but isn’t really as significant as it sounds: if it’s off, you can throw from the error handler instead, and catch at an outer scope.

Then the default does not match the blog post. The default only changes if you changed the engine exception runtime setting. The default remains a TypeError. I interpreted the blog post text as `hhvm will not throw a TypeError anymore and hard abort your request with a hard unrecoverable error.`

Reply:
Good catch, I'll clarify that.